libsel4vm/src/arch/arm/vgic/vgic_v2.c

```diff
@@ -0,0 +1,300 @@
+        ignore_fault(vcpu->vcpu_arch.fault);
+    }
+
+    // vm->unlock();
+
+    return err;
+}
+
+static memory_fault_result_t handle_vgic_vcpu_fault(vm_t *vm, vm_vcpu_t *vcpu, uintptr_t fault_addr,
+                                                    size_t fault_length,
+                                                    void *cookie)
+{
+    /* We shouldn't fault on the vgic vcpu region as it should be mapped in
+     * with all rights */
+    return FAULT_ERROR;
+}
+
+static vm_frame_t vgic_vcpu_iterator(uintptr_t addr, void *cookie)
+{
+    cspacepath_t frame;
+    vm_frame_t frame_result = { seL4_CapNull, seL4_NoRights, 0, 0 };
+    vm_t *vm = (vm_t *)cookie;
+
+    int err = vka_cspace_alloc_path(vm->vka, &frame);
+    if (err) {
+        ZF_LOGE("Failed to allocate cslot for vgic vcpu");
+        return frame_result;
+    }
+    seL4_Word vka_cookie;
+    err = vka_utspace_alloc_at(vm->vka, &frame, kobject_get_type(KOBJECT_FRAME, 12), 12, GIC_VCPU_PADDR, &vka_cookie);
+    if (err) {
+        err = simple_get_frame_cap(vm->simple, (void *)GIC_VCPU_PADDR, 12, &frame);
+        if (err) {
+            ZF_LOGE("Failed to find device cap for vgic vcpu");
+            return frame_result;
+        }
+    }
+    frame_result.cptr = frame.capPtr;
+    frame_result.rights = seL4_AllRights;
+    frame_result.vaddr = GIC_CPU_PADDR;
+    frame_result.size_bits = seL4_PageBits;
+    return frame_result;
+}
+
+/*
+ * 1) completely virtual the distributor
+ * 2) remap vcpu to cpu. Full access
+ */
+int vm_install_vgic(vm_t *vm)
+{
+    struct vgic *vgic = calloc(1, sizeof(*vgic));
+    if (!vgic) {
+        assert(!"Unable to calloc memory for VGIC");
+        return -1;
+    }
+    /* vgic doesn't require further initialization, having all fields set to
+     * zero is fine.
+     */
+
+    /* Distributor */
+    vgic_dist = (struct vgic_dist_device *)calloc(1, sizeof(struct vgic_dist_device));
+    if (!vgic_dist) {
+        return -1;
+    }
+    memcpy(vgic_dist, &dev_vgic_dist, sizeof(struct vgic_dist_device));
+
+    vgic->dist = calloc(1, sizeof(struct gic_dist_map));
+    assert(vgic->dist);
+    if (vgic->dist == NULL) {
+        return -1;
+    }
+    vm_memory_reservation_t *vgic_dist_res = vm_reserve_memory_at(vm, GIC_DIST_PADDR, PAGE_SIZE_4K,
+                                                                  handle_vgic_dist_fault, (void *)vgic_dist);
+    vgic_dist->priv = (void *)vgic;
+    vgic_dist_reset(vgic_dist);
+
+    /* Remap VCPU to CPU */
+    vm_memory_reservation_t *vgic_vcpu_reservation = vm_reserve_memory_at(vm, GIC_CPU_PADDR, PAGE_SIZE_4K,
+                                                                          handle_vgic_vcpu_fault, NULL);
+    int err = vm_map_reservation(vm, vgic_vcpu_reservation, vgic_vcpu_iterator, (void *)vm);
+    if (err) {
+        free(vgic_dist->priv);
+        return -1;
+    }
+
+    return 0;
+}
+
+int vm_vgic_maintenance_handler(vm_vcpu_t *vcpu)
```

Comment:
Same here, this function and alsdo `handle_vgic_maintenance()` are the same in GICv2 and GIV3, so I'd keep this in a common file for now.

Reply:
They're slightly different since GICv3 uses group 1 while GICv2 uses group 0.

Reply:
(only `handle_vgic_maintenance` differs)

Reply:
True, but that could be a constant then. My main point at the moment it trying to have GICv3 support with out renaming the source file and moving code, so the change are easier to maintain.